theano/tensor/nnet/nnet.py

```diff
@@ -1784,6 +1803,59 @@ def graph_merge_softmax_with_crossentropy_softmax(node):
                     return [mergeable_client[1]]
 
 
+@opt.register_specialize
+@opt.register_stabilize
+@opt.register_canonicalize
+@gof.local_optimizer([CrossentropySoftmax1HotWithBiasDx])
+def local_useless_crossentropy_softmax_1hot_with_bias_dx_alloc(node):
+    """
+    Replaces a CrossentropySoftmax1HotWithBiasDx op, whose incoming gradient is
+    an `alloc` of a scalar variable or one that has either broadcastable or
+    matching dimensions with the output variable, by one that skips the
+    intermediate `alloc`.
+    """
+    if isinstance(node.op, CrossentropySoftmax1HotWithBiasDx):
+        dy, sm, y_idx = node.inputs
+
+        if dy.owner is not None and isinstance(dy.owner.op, tensor.Alloc):
+            # dz is the input of the Alloc op, i.e. T.alloc(dz, <shape>)
+            dz = dy.owner.inputs[0]
+
+            # A non-empty list of conditions means that `dz` has some
+            # non-broadcastable dimensions that cannot be checked statically
+            # for equal size with the corresponding dimension of `sm`. Thus, a
+            # runtime check is necessary.
+            try:
+                cond = [tensor.eq(dy.shape[k], sm.shape[k])
```

Comment:
Likely, I need to look into this again after we've solved the shape inference issue. See my comment below.